### DISCUSSION TITLE
Tutorial info

### DISCUSSION THREAD

**Discussion by uvtc** [libui-ng/libui-ng#60] [General]
Overall, from the user's (application programmer's) perspective, how does libui-ng work?

What standard or non-standard terminology does libui-ng use? ("events"? "callbacks"? "bindings"? "window/panes/frames"?, "box"? "parent/child" widgets? etc.?)

How are events connected to or bound to widgets?

How are widgets configured, enabled/disabled?

How does libui-ng handle laytout / geometry? Is it anything like Tk's grid? vboxes and hboxes? Something else?

Explanations, expositions, comments, complaints, juicy morsels, random factoids, incomplete tidbits,  ... all most gratefully appreciated.

Thanks!

**Comment by cody271** [comment]
> Overall, from the user's (application programmer's) perspective, how does libui-ng work?
> 
libui-ng is a very thin wrapper over your desktop platform's native GUI framework.
This allows for an extremely minimalist API that is still quite featureful. Won't ever have the full breadth of features of something like Qt or Electron, but operates at order(s) of magnitude lower resource footprint.

Our three platforms currently are:

`darwin`: [AppKit](https://en.wikipedia.org/wiki/AppKit) on macOS, also known as "Cocoa".
`unix`: [GTK3](https://en.wikipedia.org/wiki/GTK) on any Wayland or X11 Linux desktop. X11 only on the BSDs. Soon will be GTK4 and maybe [libadwaita](https://gitlab.gnome.org/GNOME/libadwaita).
`windows`: Original [Win32](https://docs.microsoft.com/en-us/windows/win32/desktop-programming) API. Soon will probably have to be Win32/[WinRT](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/) hybrid using [XAML Islands](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands).

> What standard or non-standard terminology does libui-ng use?
> "events"? "callbacks"? "bindings"?
>
Yes, "events" are bound to "callbacks". Integrations with other languages are called "bindings".
Widgets are called "controls", and all inherit from `uiControl` base class.

> "window/panes/frames"?, "box"?
>
Yes `uiWindow` and `uiBox`. Frames control is called `uiGroup`. No "panes" yet.

> "parent/child" widgets? etc.?
>
See `uiControlParent()`/`uiControlSetParent()`.
Access to a container control's children is class specific. For example, `uiBoxNumChildren()`.

> How are events connected to or bound to widgets?
>
Each control has an API for each event to assign a callback.
For example, `uiButtonOnClicked()` to respond to a button being clicked.

> How are widgets configured?
> 
Same as callbacks, there is no unified control properties API yet.
You must use the control-specific functions. For example, `uiLabelSetText()`/`uiLabelText()` for label text.

> enabled/disabled?
>
Setters are `uiControlEnable()`/`uiControlDisable()`. Getter is `uiControlEnabled()`.

> How does libui-ng handle laytout / geometry?
> Is it anything like Tk's grid?
> 
`uiGrid` is exactly that yes.

>  vboxes and hboxes?
>
Yes, see `uiNewVerticalBox()` and `uiNewHorizontalBox()`.

>  Something else?
> 
`uiFixed` allows for integration with any external layout engine.
The main use case so far is for CSS flexbox style layout using [Yoga](https://yogalayout.com).

> Explanations, expositions, comments, complaints, juicy morsels, random factoids, incomplete tidbits, ... all most gratefully appreciated.
> 
As a Lua and occasional Clojure dev myself, Janet looks interesting!
Unlocking the wealth of tools/talent specific to these kinds of niche languages was always a design goal here.

> Thanks!
>
Hope that all makes sense and is helpful!

**Comment by uvtc** [comment]
Thanks so much, @cody271 .
